yamux/src/lib.rs

```diff
@@ -95,70 +76,89 @@ pub enum WindowUpdateMode {
 /// - split send size = 16 KiB
 #[derive(Debug, Clone)]
 pub struct Config {
-    receive_window: u32,
-    max_buffer_size: usize,
+    max_stream_receive_window: Option<u32>,
+    max_connection_receive_window: usize,
     max_num_streams: usize,
-    window_update_mode: WindowUpdateMode,
     read_after_close: bool,
     split_send_size: usize,
 }
 
 impl Default for Config {
     fn default() -> Self {
         Config {
-            receive_window: DEFAULT_CREDIT,
-            max_buffer_size: 1024 * 1024,
-            max_num_streams: 8192,
-            window_update_mode: WindowUpdateMode::OnRead,
+            // TODO: Add rational: given that we have a connection window, ...
+            max_stream_receive_window: None,
+            // TODO: reevaluate default.
+            max_connection_receive_window: 1 * 1024 * 1024 * 1024,
+            // TODO
+            max_num_streams: 512,
             read_after_close: true,
             split_send_size: DEFAULT_SPLIT_SEND_SIZE,
         }
     }
 }
 
 impl Config {
-    /// Set the receive window per stream (must be >= 256 KiB).
+    /// Set the maximum receive window per stream .
+    ///
+    /// Must be `>= 256 KiB`.
+    ///
+    /// The window of a stream starts at 256 KiB and is increased (auto-tuned) based on the
+    /// connection's round-trip time and the stream's bandwidth (striving for the
+    /// bandwidth-delay-product).
+    ///
+    /// Set to `None` to disable the per stream maximum receive window.
     ///
     /// # Panics
     ///
     /// If the given receive window is < 256 KiB.
-    pub fn set_receive_window(&mut self, n: u32) -> &mut Self {
-        assert!(n >= DEFAULT_CREDIT);
-        self.receive_window = n;
+    pub fn set_max_stream_receive_window(&mut self, n: Option<u32>) -> &mut Self {
```

Comment:
Why do we need to make this configurable? Isn't it okay for a single stream to increase its buffer as long as it stays within the configured max memory usage of the entire connection?